Implement endpoint /twitter to get twitter feed closes #65

PreetPatel opened the pull request:
GitHub Issue (If applicable): #65 

## PR Type

What kind of change does this PR introduce?
<!-- Please uncomment one or more that apply to this PR

- Bugfix
- Feature
- Code style update (formatting)
- Refactoring (no functional changes, no api changes)
- Build or CI related changes
- Documentation content changes
- Project automation
- Other... Please describe:

-->
- Feature

## What is the current behavior?

Hitting the endpoint /twitter does not return anything


## What is the new behavior?

Hitting the endpoint /twitter after adding the correct application.properties returns a JSON of the Twitter feed


## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] Tested code with current upstream repo
- [ ] Added new tests to the test suite for the feature (@EricPedrido  will make a seperate PR for this)
- [ ] Tested the full test suite, and ensured that the feature does not break current build.
- [ ] Docs have been added/updated which fit (for bug fixes / features)
- [x] Contains **NO** breaking changes
- [x] Associated with an issue (GitHub or internal)
- [x] Checked licensing of Frameworks/Libraries (if applicable) 


<!-- If this PR contains a breaking change, please describe the impact below, and why this change had to be introduced.
     Please note that breaking changes are likely to be rejected -->


## Other information

Please check Discord for the correct application.properties to run this

Internal Issue (If applicable):
<!-- Link to relevant internal issue if applicable. All PRs should be associated with an issue (GitHub issue or internal) -->
<!-- Template modified from Platform Uno Github, under the Apache 2.0 License-->

Minus20Five commented on server/src/main/java/com/feeder/server/model/TwitterData.java:
```diff
@@ -1,3 +1,56 @@
+        return Type.TWITTER;
+    }
+
+    /** This property is required for serialization. This means all posts must contain a title. */
```
Old comment

Minus20Five commented on server/src/main/java/com/feeder/server/model/TwitterData.java:
```diff
@@ -1,3 +1,56 @@
+    /**
+     * {@link java.util.Optional} means this property is NOT required for serialization. Perhaps not
+     * all posts contain images.
+     */
```
Old comment

PreetPatel commented on server/src/test/java/com/feeder/server/provider/reddit/RedditClientConfigTest.java:
```diff
@@ -13,7 +13,7 @@
   @BeforeEach
   public void setup() {
-    emptyProperties = new ApplicationProperties("", "", "", "");
+    emptyProperties = new ApplicationProperties("", "", "", "","","","","");
```
will need to do something about this.. itll keep having to be updated as more application properties are added

jordansimsmith replied:
I agree. Could you raise an issue for it @PreetPatel ?

PreetPatel replied:
Nvm.. addressed in my upcoming commit

PreetPatel commented:
Yes I know this PR doesn't have tests. @EricPedrido is adding tests so theres equal contribution :)

Minus20Five commented on server/src/main/java/com/feeder/server/provider/twitter/TwitterCredentials.java:
```diff
@@ -0,0 +1,32 @@
+        } else {
+            new TwitterCredentials(properties);
+            return twitterFactory.getInstance();
+        }
```
What happens if this was just:

```suggestion
            return twitterFactory.getInstance();
```
Also I don't understand why `TwitterCredentials` is created then discarded.

PreetPatel replied:
You need to build the factory with configurations first.. could have placed it in this method but chose to not do that because it didn't make sense to me to

Minus20Five replied:
I tested it with this change and it seems to work fine?

PreetPatel replied:
Weirdly yes.. will change

Minus20Five commented on server/src/main/java/com/feeder/server/provider/twitter/TwitterFeedProvider.java:
```diff
@@ -2,14 +2,42 @@
+    List<TwitterData> modelledResponse = new ArrayList<>();
+
+    try {
+      statuses = twitter.getHomeTimeline();
```
Doing this blocks the request handler thread. It would be better if this was:
`return Flux.fromIterable(twitter.getHomeTimeline()).flatMap(status -> TwitterData.newBuilder().....)`

PreetPatel replied:
ah didnt know that

jordansimsmith commented on server/src/main/java/com/feeder/server/provider/twitter/TwitterCredentials.java:
```diff
@@ -0,0 +1,32 @@
+import twitter4j.conf.ConfigurationBuilder;
+
+@Configuration
+class TwitterCredentials {
```
Should probably have a visibility modifier here

PreetPatel replied:
The idea was to keep this class package-private to just inside the twitter package.. hence the missing modifier to keep it package-private

jordansimsmith replied:
That makes sense. Maybe a comment to explain that would be helpful for new viewers :)

jordansimsmith commented on server/src/main/java/com/feeder/server/model/TwitterData.java:
```diff
@@ -1,3 +1,50 @@
+    }
+
+    @JsonProperty("tweet")
+    public abstract String tweet();
```
@PreetPatel is it also possible to get the image from the tweet body? If not, no worries.

PreetPatel replied:
Twitter doesn't provide that in its API. It just appends a link to twitter.com in the body of all of its statuses

HarrisonLeach1 commented on server/src/main/java/com/feeder/server/provider/twitter/TwitterCredentials.java:
```diff
@@ -0,0 +1,32 @@
+    TwitterCredentials(ApplicationProperties properties) {
+        ConfigurationBuilder cb = new ConfigurationBuilder();
+        cb.setDebugEnabled(true)
+                .setOAuthConsumerKey(properties.getTwitterConsumerKey())
```
Not essential, but if a new developer were to forget to set their twitter secrets they would be confronted by this error message: 

![tweet errror](https://user-images.githubusercontent.com/44953072/76691395-3f910500-66af-11ea-8174-1fae355d1188.png)

This could be confusing for a first-time contributor and may take them a while to debug.

One option to address this is to validate that the twitter secrets aren't empty first, if they are you could provide the following error message instead:
`providers.twitter.consumerKey in application.properties is empty`

PreetPatel replied:
Think in the most earliest log (if you scroll a bit up), Twitter responds with a 404 and a status body saying "This may be caused if no keys are provided"

Minus20Five commented:
Hi I would like to propose two changes.
1) Twitter has quite aggressive [rate limiting](https://developer.twitter.com/en/docs/basics/rate-limiting) (seems to be 15 calls every 15 minutes). So we should cache the response in case we get limited.

2) Lets move the blocking call to a different thread pool

jordansimsmith commented:
> Hi I would like to propose two changes.
> 
> 1. Twitter has quite aggressive [rate limiting](https://developer.twitter.com/en/docs/basics/rate-limiting) (seems to be 15 calls every 15 minutes). So we should cache the response in case we get limited.
> 2. Lets move the blocking call to a different thread pool

I agree with this, however, I'd like to split this to another issue as this PR is blocking frontend integration.

PreetPatel commented on server/src/main/java/com/feeder/server/provider/twitter/TwitterCredentials.java:
```diff
@@ -0,0 +1,32 @@
+@Configuration
+class TwitterCredentials {
+    private TwitterFactory twitterFactory;
+    TwitterCredentials(ApplicationProperties properties) {
```
Addressed in the comment above https://github.com/softeng-701-group-5/softeng-701-assignment-1/pull/71#discussion_r392629850

PreetPatel commented:
> > Hi I would like to propose two changes.
> > 
> > 1. Twitter has quite aggressive [rate limiting](https://developer.twitter.com/en/docs/basics/rate-limiting) (seems to be 15 calls every 15 minutes). So we should cache the response in case we get limited.
> > 2. Lets move the blocking call to a different thread pool
> 
> I agree with this, however, I'd like to split this to another issue as this PR is blocking frontend integration.

Agree.. This can be for some other team to look at

Minus20Five commented:
I have implemented these two changes but I can't push them.

The rate limiting issue is a must-fix because it will crash the server if you refresh the app more than 15 times / 15 minutes.

harmanlamba commented:
> I have implemented these two changes but I can't push them.
> 
> The rate limiting issue is a must-fix because it will crash the server if you refresh the app more than 15 times / 15 minutes.

Hey @Minus20Five , did you implement them on this branch itself? If that's the case @PreetPatel , could you please give @Minus20Five push access to your fork, and we can close another issue with this PR? Great work otherwise Twitter team! Looking forward to having this integrated into the front-end!

PreetPatel commented:
> > I have implemented these two changes but I can't push them.
> > The rate limiting issue is a must-fix because it will crash the server if you refresh the app more than 15 times / 15 minutes.
> 
> Hey @Minus20Five , did you implement them on this branch itself? If that's the case @PreetPatel , could you please give @Minus20Five push access to your fork, and we can close another issue with this PR? Great work otherwise Twitter team! Looking forward to having this integrated into the front-end!

Waiting for 15 mins to go by so I can test what I just implemented for caching.. will give the all clear then
